proxy/proxy.go

```diff
@@ -315,17 +325,34 @@ func (p *Proxy) handle(conn net.Conn) {
 		ctx:                 p.ctx,
 		proxy:               p,
 		preparedSystemQuery: make(map[[preparedIdSize]byte]interface{}),
+		codec:               codecs.CustomRawCodec,
 	}
 	p.addClient(cl)
 	cl.conn = proxycore.NewConn(conn, cl)
 	cl.conn.Start()
 }
 
-func (p *Proxy) maybeCreateSession(version primitive.ProtocolVersion, keyspace string) (*proxycore.Session, error) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	if cachedSession, ok := p.sessions[version].Load(keyspace); ok {
-		return cachedSession.(*proxycore.Session), nil
+func (p *Proxy) maybeCreateSession(version primitive.ProtocolVersion, keyspace, compression string) (*proxycore.Session, error) {
+	p.sessionsMu.RLock()
+	defer p.sessionsMu.RUnlock()
+	return p.maybeCreateSessionUnlocked(version, keyspace, compression)
+}
+
+func (p *Proxy) findSession(version primitive.ProtocolVersion, keyspace, compression string) (*proxycore.Session, error) {
+	p.sessionsMu.RLock()
+	defer p.sessionsMu.RUnlock()
+	key := sessionKey{version: version, keyspace: keyspace, compression: compression}
+	if s, ok := p.sessions[key]; ok {
+		return s, nil
+	} else {
+		return p.maybeCreateSessionUnlocked(version, keyspace, compression)
+	}
+}
+
+func (p *Proxy) maybeCreateSessionUnlocked(version primitive.ProtocolVersion, keyspace, compression string) (*proxycore.Session, error) {
+	key := sessionKey{version: version, keyspace: keyspace, compression: compression}
+	if cachedSession, ok := p.sessions[key]; ok {
+		return cachedSession, nil
```

Comment:
wrong use of RWMutex: `maybeCreateSessionUnlocked` now has a concurrent map assign

Reply:
@mpenick

Reply:
I put up a fix: https://github.com/datastax/cql-proxy/pull/154 @absurdfarce

Reply:
Agreed with @mpenick... thanks for the report @moguchev!

Fix has been reviewed, approved and merged.